funsor/gaussian.py

```diff
@@ -0,0 +1,263 @@
+        nonreal_inputs = OrderedDict((k, d) for k, d in lhs.inputs.items()
+                                     if d.dtype != 'real')
+        inputs, (rhs_data, log_density, loc, scale_tril) = align_tensors(
+                rhs,
+                Tensor(lhs.log_density, nonreal_inputs),
+                Tensor(lhs.loc, nonreal_inputs),
+                Tensor(lhs.scale_tril, nonreal_inputs))
+        log_density = op(log_density, rhs_data)
+        inputs.update(lhs.inputs)
+        return Gaussian(log_density, loc, scale_tril, inputs)
+
+    if op is ops.mul or op is ops.truediv:
+        # Scale a Gaussian, as under pyro.poutine.scale.
+        raise NotImplementedError('TODO')
+
+    return None  # defer to default implementation
+
+
+@eager.register(Binary, object, Tensor, Gaussian)
+def eager_binary_tensor_gaussian(op, lhs, rhs):
+    if op is ops.add:
+        # Add a batch-dependent log_density term to a Gaussian.
+        nonreal_inputs = OrderedDict((k, d) for k, d in rhs.inputs.items()
+                                     if d.dtype != 'real')
+        inputs, (lhs_data, log_density, loc, scale_tril) = align_tensors(
+                lhs,
+                Tensor(rhs.log_density, nonreal_inputs),
+                Tensor(rhs.loc, nonreal_inputs),
+                Tensor(rhs.scale_tril, nonreal_inputs))
+        log_density = op(lhs_data, log_density)
+        inputs.update(rhs.inputs)
+        return Gaussian(log_density, loc, scale_tril, inputs)
+
+    if op is ops.mul:
+        # Scale a Gaussian, as under pyro.poutine.scale.
+        raise NotImplementedError('TODO')
+
+    return None  # defer to default implementation
+
+
+@eager.register(Binary, object, Gaussian, Gaussian)
+def eager_binary_gaussian_gaussian(op, lhs, rhs):
+    if op is ops.add:
+        # Fuse two Gaussians by adding their log-densities pointwise.
+        # This is similar to a Kalman filter update, but also keeps track of
+        # the marginal likelihood.
+
+        # Align moments.
+        lhs_precision = _scale_tril_to_precision(lhs.scale_tril)
+        rhs_precision = _scale_tril_to_precision(rhs.scale_tril)
+        inputs, lhs_moments, rhs_moments = align_gaussians(
+            lhs.inputs, lhs.log_density, lhs.loc, lhs_precision,
+            rhs.inputs, rhs.log_density, rhs.loc, rhs_precision)
+        lhs_log_density, lhs_loc, lhs_precision = lhs_moments
+        rhs_log_density, rhs_loc, rhs_precision = rhs_moments
+
+        # Fuse aligned Gaussians.
+        precision = lhs_precision + rhs_precision
+        scale_tril_inv = torch.cholesky(precision)
+        scale_tril = torch.inverse(scale_tril_inv)
+        precision_loc = _mv(lhs_precision, lhs_loc) + _mv(rhs_precision, rhs_loc)
+        loc = _mv(scale_tril, _mv(scale_tril.transpose(-1, 2), precision_loc))
+        log_density = (lhs_log_density + rhs_log_density +  # FIXME add missing terms
+                       _batch_mahalanobis(scale_tril, lhs_loc - rhs_loc))
```

Comment:
It seems to me (see [my derivation](https://gist.github.com/fehiepsi/5ef8e09e61604f10607380467eb82006)) that the required term is
```
_batch_mahalanobis(lhs.scale_tril, lhs_loc) + _batch_mahalanobis(rhs.scale_tril, rhs.scale_tril) - _batch_mahalanobis(scale_tril, loc) + (logdet_terms if it is involved??)
```
I am not sure I have derived the terms in a correct way. As a side node, we can use Woobury matrix identity to compute `scale_tril` (to avoid working with precision matrix which involves `inverse` at many places). But we can do it later (I'll update the above gist in the mean time).

Reply:
Thanks @fehiepsi! Yes I'd like to improve the math after this initial PR.

Reply:
@fehiepsi I believe the formula in your derivation is correct, but I believe a more numerically stable version can be arrived at by restricting to a delta term `lhs_loc - rhs_loc`, I believe something like
```
(lhs_loc - rhs_loc)' * ((lhs_precision rhs_precision) / precision) * (lhs_loc - rhs_loc)
```
with suitable realignments. I had derived this for the univariate case, but I'm not sure how to adapt it to multivariate.

Reply:
Thanks @fritzo ! I agree that we can cancel some terms involving `c` in the notebook to arrive at your formula with `p_1 p_2 / (p_1 + p_2) = `(cov_1 + cov_2)^{-1}`. I'll work that out tomorrow, including the logdet terms.

Reply:
@fritzo I just updated the notebook with come up with the multivariate form of your formula. It is interesting that we can derive a nice form for log density adjustment (it is log density at lhs_loc - rhs_loc of a Gaussian with covariance is the sum of two covariances). Though I still think that using the formula involving Mahalanobis terms is better, we should address it later because this PR is already big!

Please ping me when this PR is ready so I'll take another pass for other places too. :)